### PR TITLE
feat: pagination + jwt authentication for get reviews API

### PR DESCRIPTION
**Pull request by chinathaip**
## What have I done?
the get reviews API now uses JWT auth middleware to determine the number of reviews users can retrieve

if the user is anonymous (no token provided in the request header) or never had their reviews approved --> can fetch only 2 reviews per course

if the user has written a review and it has been approved --> we will use the `pageSize` and `pageNumber` from the request for pagination (frontend should limit pageSize to 10 per 1 api call)

## Usages
- without any query parameters (will use default value of pageSize=2 and pageNumber=1)
`GET /api/courses/:courseCode/reviews`

#### Example:
`GET http://localhost:8003/api/courses/PHYS101/reviews`

--------------------------------------------------------------

- with pagination parameters
`GET /api/courses/:courseCode/reviews?pageSize=...&pageNumber...`

     - `pageSize` determines how many items to retrieve in 1 api call

     - `pageNumber` determines the offset
#### Example:

`GET http://localhost:8003/api/courses/PHYS101/reviews?pageSize=10&pageNumber=1`

--------------------------------------------------------------

**Request Headers**
Authorization - Bearer token (should be generated from Supabase)

Example: 
```eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImtoaW5nQHN0dWRlbnRzLnN0YW1mb3JkLmVkdSIsImV4cCI6MTcwMzY5NzA2Miwic3ViIjoiOGE3YjNjMmUtM2U1Zi00ZjFhLWE4YjctM2MyZTFhNGY1YjZkIn0.4S2QplWtjfoNe9v8cCpjWrl77Dsvjd4TQ9mJghWmJtg```

Response:
```json
{
    "pageInformation": {
        "number": 1,
        "size": 10
    },
    "totalNumberOfItems": 4,
    "totalPages": 1,
    "data": [
        {
            "id": 7,
            "academicYear": 2023,
            "description": "I hate this",
            "isOwner": true,
            "rating": 2,
            "status": "APPROVED",
            "votes": 0,
            "course": {
                "id": 3,
                "code": "PHYS101"
            },
            "profile": {
                "id": "8a7b3c2e-3e5f-4f1a-a8b7-3c2e1a4f5b6d"
            }
        },
        {
            "id": 8,
            "academicYear": 2021,
            "description": "I don't really care",
            "rating": 2,
            "status": "APPROVED",
            "votes": 2,
            "course": {
                "id": 3,
                "code": "PHYS101"
            },
            "profile": {
                "id": "bae5c487-96e8-4c93-b831-508a29d6d887"
            }
        },
        {
            "id": 3,
            "academicYear": 2022,
            "description": "Not a fan of the teaching style.",
            "rating": 2,
            "status": "APPROVED",
            "votes": 5,
            "course": {
                "id": 3,
                "code": "PHYS101"
            },
            "profile": {
                "id": "2d1f3c4e-5a6b-7c8d-9e0f-1a2b3c4d5e6f"
            }
        },
        {
            "id": 2,
            "academicYear": 2022,
            "description": "The material was challenging but interesting.",
            "rating": 4,
            "status": "APPROVED",
            "votes": 8,
            "course": {
                "id": 3,
                "code": "PHYS101"
            },
            "profile": {
                "id": "5c4b3c4d-5e6f-7a8b-9c0d-1e2f3a4b5c6d"
            }
        }
    ]
}
```

## Tests?
- pagination_test.go has test coverage of 90.9% (actually it should be 100 but go test is kinda buggy)
- get_review.go has test coverage of  91.1%
image
image

### CODE REVIEW

File: apps/reviews-api/data/get_review.go [hunk 1]
```diff
@@ -2,48 +2,149 @@ package data
...
+		pageNumber = 1
+		totalNumberOfItems = maxReviewsForNonActiveUsers
+	}
+
+	myReview := <-myReviewChan
+	if myReview != nil {
+		totalNumberOfItems++
+	}
+
+	start, end := utils.CalculateStartEnd(rawReviews, pageSize, pageNumber)
```
Comment (Lxkas) [inline apps/reviews-api/data/get_review.go:142]: Couldn't we just use LIMIT and OFFSET in the query?

**Review by Lxkas — CHANGES_REQUESTED** [review]

File: apps/reviews-api/data/get_review.go (hunk 1 above)
Comment (Lxkas) [reply apps/reviews-api/data/get_review.go:142]: I meant to press comment but GitHub :shrug:

Comment (chinathaip) [reply apps/reviews-api/data/get_review.go:142]: > Couldn't we just use LIMIT and OFFSET in the query?

I retrieved the whole thing to determine the total number of pages that can be sent along the response.

This way the frontend can do stuff like showing the total number of pages at the bottom, or know whether or not they can fetch more data

Using limit and offset might not allow this idea (or is there any other way to do it?)

Comment (Lxkas) [reply apps/reviews-api/data/get_review.go:142]: Prolly better to just have a function like QueryTotalCountOfReviews and call that and return it along with the result, something like `SELECT COUNT(*)` using COUNT is much faster than querying every row, and the original query can be made into LIMIT + OFFSET

Comment (chinathaip) [reply apps/reviews-api/data/get_review.go:142]: okie dokie

prisma go apparently does not support COUNT, so imma go with raw query XD

Comment (Lxkas) [reply apps/reviews-api/data/get_review.go:142]: > okie dokie
> 
> prisma go apparently does not support COUNT, so imma go with raw query XD

Wowwww no COUNT support?! Go kinda laaaame~? 😃

Comment (chinathaip) [reply apps/reviews-api/data/get_review.go:142]: THE LIB IS ACTUALLY ASS!!!

**Review by Lxkas — APPROVED** [review]
Okie dokie brokie hokie

**Pull request merged**